src/server/io_utils.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/io_utils.h"
+
+#include "base/flags.h"
+#include "base/logging.h"
+#include "server/error.h"
+
+namespace dfly {
+
+void BufferedStreamerBase::ReportWritten() {
+  buffered_++;
+  // Check and possibly wait for consumer to keep up.
+  waker_.await([this]() { return !IsStalled() || IsStopped(); });
+  // Notify consumer we have new data.
+  waker_.notify();
+}
+
+std::error_code BufferedStreamerBase::WriteAll(io::Sink* dest) {
+  CHECK_NE(buf_ptr_, nullptr);
+
+  while (!IsStopped()) {
+    // Wait for more data or stop signal.
+    waker_.await([this]() { return buffered_ > 0 || IsStopped(); });
+    if (IsStopped())
+      break;
+
+    // Steal-swap producer buffer.
+    std::swap(buf_stash_, *buf_ptr_);
```

Comment:
I think that the flow of switching the buffer of journal writer without it knowing is not very intuitive. 
How about creating journal writer with buffer not a member but one that is set from outside, with this approach you will have SetBuffer func to replace the buf and you can use it here.

Reply:
Otherwise, we can only keep the Flush(sink) part and flush to an IoBuf which will be owned by the buffered stream

Reply:
@romange What do you think?

Reply:
it's localized and does not affect lots of places in the code.
and we can make it more clear by adding more comments.